File: baserecycleradapter/src/main/java/com/gigigo/baserecycleradapter/viewholder/BaseViewHolder.java
package com.gigigo.baserecycleradapter.viewholder;

import android.content.Context;
import android.view.DragEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

public abstract class BaseViewHolder<V> extends RecyclerView.ViewHolder
        implements View.OnLongClickListener, View.OnClickListener, View.OnDragListener {

    private OnItemClickListener itemClickListener;
    private OnItemLongClickListener itemLongClickListener;
    private OnItemDragListener itemDragListener;

    public BaseViewHolder(Context context, ViewGroup parent, int layoutId) {
        super(LayoutInflater.from(context).inflate(layoutId, parent, false));
        bindListeners();
    }

    public BaseViewHolder(View itemView) {
        super(itemView);
    }

    private void bindListeners() {
        itemView.setOnClickListener(this);
        itemView.setOnLongClickListener(this);
        itemView.setOnDragListener(this);
    }

    public void setItemClickListener(OnItemClickListener itemClickListener) {
        this.itemClickListener = itemClickListener;
    }

    public void setItemLongClickListener(OnItemLongClickListener itemLongClickListener) {
        this.itemLongClickListener = itemLongClickListener;
    }

    public void setItemDragListener(OnItemDragListener itemDragClickListener) {
        this.itemDragListener = itemDragClickListener;
    }

    @Override
    public void onClick(View v) {
        if (itemClickListener == null) return;
        itemClickListener.onItemClick(getLayoutPosition(), v);
    }

    @Override
    public boolean onLongClick(View v) {
        if (itemLongClickListener == null) {
            return false;
        }
        return itemLongClickListener.onItemLongClicked(getLayoutPosition(), itemView);
    }

    @Override
    public boolean onDrag(View v, DragEvent event) {
        if (itemDragListener == null) {
            return false;
        }
        return itemDragListener.OnItemDragged(getLayoutPosition(), v);
    }

    public abstract void bindTo(V value, int position);

    public interface OnItemClickListener {
        void onItemClick(final int position, View view);
    }

    public interface OnItemLongClickListener {
        boolean onItemLongClicked(final int position, View view);
    }

    public interface OnItemDragListener {
        boolean OnItemDragged(final int position, View view);
    }
}

File: baserecycleradapter/src/main/java/com/gigigo/baserecycleradapter/viewholder/BaseViewHolder.kt
package com.gigigo.baserecycleradapter.viewholder

import android.content.Context
import android.view.DragEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import androidx.recyclerview.widget.RecyclerView

typealias OnItemClickListener = (Int, View) -> Unit
typealias OnItemLongClickListener = (Int, View) -> Boolean
typealias OnItemDragListener = (Int, View) -> Boolean

abstract class BaseViewHolder<Data> : RecyclerView.ViewHolder, View.OnLongClickListener,
    View.OnClickListener, View.OnDragListener {

    private var itemClickListener: OnItemClickListener? = null
    private var itemLongClickListener: OnItemLongClickListener? = null
    private var itemDragListener: OnItemDragListener? = null

    constructor(context: Context, parent: ViewGroup, layoutId: Int) : super(
        LayoutInflater.from(
            context
        ).inflate(layoutId, parent, false)
    ) {
        bindListeners()
    }

    constructor(itemView: View) : super(itemView) {}

    private fun bindListeners() {
        itemView.setOnClickListener(this)
        itemView.setOnLongClickListener(this)
        itemView.setOnDragListener(this)
    }

    fun setItemClickListener(itemClickListener: OnItemClickListener?) {
        this.itemClickListener = itemClickListener
    }

    fun setItemLongClickListener(itemLongClickListener: OnItemLongClickListener?) {
        this.itemLongClickListener = itemLongClickListener
    }

    fun setItemDragListener(itemDragClickListener: OnItemDragListener?) {
        this.itemDragListener = itemDragClickListener
    }

    override fun onClick(v: View) {
        itemClickListener?.invoke(layoutPosition, v)
    }

    override fun onLongClick(v: View): Boolean {
        return itemLongClickListener?.let { longClickListener ->
            longClickListener.invoke(layoutPosition, itemView)
            true
        } ?: false
    }

    override fun onDrag(v: View, event: DragEvent): Boolean {
        return itemDragListener?.let { dragListener ->
            dragListener.invoke(layoutPosition, v)
            true
        } ?: false
    }

    abstract fun bindTo(value: Data, position: Int)
}